flash/data/process.py

```diff
@@ -101,6 +101,151 @@ class PreprocessState:
+
+                * Output: Return a list of image paths and their associated labels.
+
+        - ``load_sample``: Function to load a sample from metadata sample.
+            Example::
+
+                * Input: Receive an image path and its label.
+
+                * Action: Load a PIL Image from received image_path.
+
+                * Output: Return the PIL Image and its label.
+
+        - ``pre_tensor_transform``: Preforms transforms on a single data sample object.
+            Example::
+
+                * Input: Receive a PIL Image and its label.
+
+                * Action: Rotate the PIL Image.
+
+                * Output: Return the rotated PIL image and its label.
+
+        - ``to_tensor_transform``: Converts a single data sample to a tensor / data structure containing tensors.
+            Example::
+
+                * Input: Receive the rotated PIL Image and its label.
+
+                * Action: Convert the rotated PIL Image to a tensor.
+
+                * Output: Return the tensored image and its label.
+
+        - ``post_tensor_transform``: Preforms transform on a single tensor sample.
+            Example::
+
+                * Input: Receive the tensored image and its label.
+
+                * Action: Flip the tensored image randomly.
+
+                * Output: Return the tensored image and its label.
+
+        - ``per_batch_transform``: Preforms transforms on a batch.
+            In this example, we decided not to override the hook.
+
+        - ``per_sample_transform_on_device``: Preforms transform on a sample already on a ``GPU`` or ``TPU``.
+            Example::
+
+                * Input: Receive a tensored image on device and its label.
+
+                * Action: Apply random transforms.
+
+                * Output: Return an augmented tensored image on device and its label.
+
+        - ``collate``: Converts a sequence of data samples into a batch.
+            Example::
+
+                * Input: Receive a list of augmented tensored images and their respective labels.
+
+                * Action: Collate the list of images into batch.
+
+                * Output: Return a batch of images and their labels.
+
+        - ``per_batch_transform_on_device``: Preforms transform a batch already on ``GPU`` or ``TPU``.
+            Example::
+
+                * Input: Receive a batch of images and their labels.
+
+                * Action: Apply normalization on the batch by substracting the mean
+                    and dividing by the standard deviation from Imagenet.
+
+                * Output: Return a normalized augmented batch of images and their labels.
+
+    .. note::
+
+        By default, each hook will be no-op. To customize them,
+        just override the hooks and ``Flash`` will take care of calling them at the right moment.
+
+    .. note::
+
+        The ``per_sample_transform_on_device`` and ``per_batch_transform`` are mutually exclusive
+        as it will impact performances.
```

Comment:
Didn't know about this. Why?

I don't think we should forbid it. Just maybe throw a warning if the perf hit is bad

Reply:
We can't yet. We will need an uncollate function.

Reply:
And re-design some of the DataPipeline internals.

Maybe in another PR.